### ISSUE TITLE
VM: Ranges and Bool operators (unexpected behaviour)

### ISSUE THREAD

**Issue by wkaluza** [fetchai/ledger#1033]
While using the Etch Playground, I noticed two cases of behaviour that surprised me as a programmer. It may be that they are intentional, but in that case they should be documented.

* Ranges are closed: the range in the Hello World example in Playground `0:5` represents `[0, 5]`, i.e. `0, 1, 2, 3, 4, 5`. I would expect a range to be semiopen, i.e. to exclude its last element. This would also give us the nice property of empty ranges being represented by `[X, X]`, which is responsible for issue #911 

* The `||` and `&&` operators are not short-circuiting. For the following code, I would expect the output to be `1 3`, but it is actually `1 2 3`.

```
function foo() : Bool
  Print('1');
  return true;
endfunction

function bar() : Bool
  Print('2');
  return false;
endfunction

function main()
    if (foo() || bar())
       Print('3');
    endif
endfunction
```

**Comment by ejfitzgerald** [comment]
This is also tracked in another issue somewhere

**Comment by bipll** [comment]
@ejfitzgerald Should they be linked then or de-duplicated?

**Comment by katharine-murphy** [comment]
Range inclusiveness and short circuiting implementation are both now documented for 0.4.x.

**Comment by wkaluza** [comment]
Fixed in b3afd9cacc60d906749ea0f404fa7cf727197533 and a02e957b033fb48f4442097d0b6df8a1a4bdb428
